### ISSUE TITLE
divide by zero handler

### ISSUE THREAD

**Issue by hellerve** [ghcjs/ghcjs#375]
Hey,

Context:
I am working on a little programming language in Haskell. I cross-compiled it using ghcjs and (almost) everything works as expected. The only thing that seems to differ in behaviour is the exception handler hooked into the REPL. It looks like this:
```
handler msg@(SomeException _) = return $ "Caught error: " ++ show (msg::SomeException)
```
and catches every exception when compiled for desktop, including divides by zero.

However, when i try to reproduce this (the divide by zero) within the web REPL, the error is propagated and the REPL dies.

How can I best resolve this? Is this a bug?

Cheers

**Comment by luite** [comment]
Yes, this should be handled the same way as with GHC, except when the exception happens in an unsafe foreign import.

**Comment by hellerve** [comment]
Hmn... It happens within my code, so this is somewhat troubling.

**Comment by luite** [comment]
do you have a small test case that reproduces this? (or a large one if not)

**Comment by hellerve** [comment]
Not really, but I can come up with one.

**Comment by luite** [comment]
ok, that'd would be great. I'm just trying to make sure that I don't miss anything important wrapping up the improved-base release

**Comment by hellerve** [comment]
Something like this should be enough:

```haskell
test = catch divideByZero handler
handler (SomeException _) = return True

divideByZero = do let x = 5 / 0
                  return True
```

If the error isn't caught (which it should), it will throw an Exception. Otherwise, `True ` is returned.

I have no idea about your setup and have no time during the week to look into it, I fear.
Does this help, though?

**Comment by luite** [comment]
I can't reproduce it with this, the `5 / 0` is ambiguous (and never forced), if I add `evaluate` and force it to be `Rational`, the exception is caught correctly in my installation (if it's `Double` there shouldn't be an exception at all, the result is NaN, which also seems to work)

If you keep seeing this problem, can you provide a running example when you have more time? It can always be fixed in a later update if needs be.

**Comment by hellerve** [comment]
Okay, sorry it didn't work.

I hope it is not just a version problem (I am using GHC 7.10.1 and ghcjs 0.1.0).

Until I get around actually doing it, [that](http://zepto.veitheller.de/try) is the project where I encounter the error and the reason why it is hard to single out the problematic bits of code: it's an interpreter for a Lisp-like language.

For fun crashes, try typing in `(/ 2 0)`. It will die (keep displaying the last evaluation frame, likely a parse error on incomplete input), whereas the native REPL just prints the exception.

**Comment by luite** [comment]
Is there a way to make the repl print the result of `(/ 2 0)` to stdout? then I can compile the original zepto interpreter to JS and run it with node, which makes debugging a bit easier.

**Comment by hellerve** [comment]
I am not completely sure what you mean by that, so I will just list everything that could be of use. 
The relevant functions in executing zepto code is `evalString(s)` in src/Zepto/Primitives.hs. The crux is that in the js-version, the error handling is baked into the main function via a where clause, like so: 
```
    where
      evaluation env x = catch (evalStrings env x) handler
      handler msg@(SomeException _) = return $ "Caught error: " ++ show (msg::SomeException)
```
I don't want you to have to go through the pain of rewriting the interpreter to make for a decent testing environment for you, though.

On the zepto side of things, you can write things to stdout via `(write "thing")` btw. At that point, the thing that should be written is evaluated though (and the error thrown), so it is likely useless.

**Comment by luite** [comment]
Oh I just mean something where I can run

```
$ zepto somefile.zp
```

in the terminal where it would print the result of `(/ 2 0)`. what would `somefile.zp` look like?

**Comment by hellerve** [comment]
Oh, you can get the interpreter [here](https://github.com/zepto-lang/zepto).

In REPL mode (which the port to JS is in), the workflow would look like this:
```
$ zepto
[...]
zepto> (/ 2 0)
Caught error: divide by zero
zepto>
```

In script mode it would die, meaning that calling this:
```clojure
(write (/ 2 0))
```
would look like this:
```
$ zepto somefile.zp
zepto: divide by zero
```

But, as the Javascript port is based on the REPL, the former is more interesting.
